sotodlib/site_pipeline/update_det_cal.py

```diff
@@ -0,0 +1,838 @@
+        if param_correction_config is None:
+            self.param_correction_config = tpc.AnalysisCfg(**kw)  # type: ignore
+        elif isinstance(param_correction_config, dict):
+            kw.update(param_correction_config)
+            self.param_correction_config = tpc.AnalysisCfg(**kw)  # type: ignore
+        else:
+            self.param_correction_config = param_correction_config
+
+        self.setup_files()
+
+    @classmethod
+    def from_yaml(cls, path) -> "DetCalCfg":
+        with open(path, "r") as f:
+            d = yaml.safe_load(f)
+            return cls(**d)
+
+    def setup_files(self) -> None:
+        """Create directories and databases if they don't exist"""
+        if not os.path.exists(self.failed_cache_file):
+            # If file doesn't exist yet, just create an empty one
+            with open(self.failed_cache_file, "w") as f:
+                yaml.dump({}, f)
+
+        if not os.path.exists(self.index_path):
+            scheme = core.metadata.ManifestScheme()
+            scheme.add_exact_match("obs:obs_id")
+            scheme.add_data_field("dataset")
+            db = core.metadata.ManifestDb(scheme=scheme)
+            db.to_file(self.index_path)
+
+
+@dataclass
+class CalInfo:
+    """
+    Class that contains detector calibration information that will go into
+    the caldb.
+
+    Attributes
+    ----------
+    readout_id: str
+        Readout ID of the detector.
+    r_tes: float
+        Detector resistance [ohms], determined through bias steps while the
+        detector is biased.
+    r_frac: float
+        Fractional resistance of TES, given by r_tes / r_n.
+    p_bias: float
+        Bias power on the TES [J] computed using bias steps at the bias point.
+    s_i: float
+        Current responsivity of the TES [1/V] computed using bias steps at the
+        bias point.
+    phase_to_pW: float
+        Phase to power conversion factor [pW/rad] computed using s_i,
+        pA_per_phi0, and detector polarity.
+    v_bias: float
+        Commanded bias voltage [V] on the bias line of the detector for the observation.
+    tau_eff: float
+        Effective thermal time constant [sec] of the detector, measured from bias steps.
+    loopgain: float
+        Loopgain of the detector.
+    tes_param_correction_success: bool
+        True if TES parameter corrections were successfully applied.
+    bg: int
+        Bias group of the detector. Taken from IV curve data, which contains
+        bgmap data taken immediately prior to IV. This will be -1 if the
+        detector is unassigned.
+    polarity: int
+        Polarity of the detector response for a positive change in bias current
+        while the detector is superconducting. This is needed to correct for
+        detectors that have reversed response.
+    r_n: float
+        Normal resistance of the TES [Ohms] calculated from IV curve data.
+    p_sat: float
+        "saturation power" of the TES [J] calculated from IV curve data.
```

Comment:
Aren't Psat / Pbias usually in W or pW?

Reply:
Yep! Thanks for catching, the stored values should be watts.